libcudacxx/include/cuda/std/detail/libcxx/include/__functional/invoke.h

```diff
@@ -481,11 +481,22 @@ struct __invoke_of
 {
 #if defined(__NVCC__) && defined(__CUDACC_EXTENDED_LAMBDA__) && \
    !defined(__CUDA_ARCH__)
-  static_assert(!__nv_is_extended_device_lambda_closure_type(_Fp),
-                "Attempt to use an extended __device__ lambda in a context "
-                "that requires querying its return type in host code. Use a "
-                "named function object, a __host__ __device__ lambda, or "
-                "cuda::proclaim_return_type instead.");
+  #if defined(CUDART_VERSION) && (CUDART_VERSION >= 12030) 
+    static_assert(!__nv_is_extended_device_lambda_closure_type(_Fp) ||
+                   __nv_is_extended_host_device_lambda_closure_type(_Fp) ||
+                   __nv_is_extended_device_lambda_with_preserved_return_type(_Fp),
+                   "Attempt to use an extended __device__ lambda in a context "
+                   "that requires querying its return type in host code. Use a "
+                   "named function object, a __host__ __device__ lambda, or "
+                   "cuda::proclaim_return_type or a extended __device__ lambda "
+                   "with preserved return type instead.");
+  #elif
```

Comment:
There is no condition here, so this should be
```suggestion
  #else // ^^^ _LIBCUDACXX_CUDACC_BELOW_12_3 ^^^ / vvv !_LIBCUDACXX_CUDACC_BELOW_12_3 vvv
```

Note that for nontrivial blocks that span multiple lines we add some markers to exactly tell the user what they should expect where

Reply:
Fixed